protocols/imap/src/main/java/org/apache/james/imap/encode/main/DefaultImapEncoderFactory.java

```diff
@@ -48,10 +57,36 @@
 import org.apache.james.imap.encode.XListResponseEncoder;
 import org.apache.james.imap.encode.base.EndImapEncoder;
 
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
 /**
  * TODO: perhaps a POJO would be better
  */
 public class DefaultImapEncoderFactory implements ImapEncoderFactory {
+    private static class DefaultImapEncoder implements ImapEncoder {
+        private final Map<Class<? extends ImapMessage>, ImapResponseEncoder> encoders;
+        private final EndImapEncoder endImapEncoder;
+
+        private DefaultImapEncoder(List<ImapResponseEncoder> encoders, EndImapEncoder endImapEncoder) {
+            this.encoders = encoders.stream()
+                .collect(Guavate.toImmutableMap(
+                    ImapResponseEncoder::acceptableMessages,
+                    Function.identity()));
+            this.endImapEncoder = endImapEncoder;
+        }
+
+        @Override
+        public void encode(ImapMessage message, ImapResponseComposer composer, ImapSession session) throws IOException {
+            ImapResponseEncoder imapResponseEncoder = encoders.get(message.getClass());
+
+            if (imapResponseEncoder != null) {
```

Comment:
did you wrote a test for unknown message type?

Reply:
Yes ^^